cmd/config-reloader/main.go

```diff
@@ -110,6 +136,26 @@ func main() {
 		serverErr <- server.ListenAndServe()
 	}()
 
+	// We want to create an output file regardless of whether we startup successfully.
+	listenAddr, err := net.ResolveTCPAddr("tcp", *listenAddress)
+	if err != nil {
+		//nolint:errcheck
+		level.Error(logger).Log("msg", "parsing listen URL failed", "err", err)
+		os.Exit(1)
+	}
+	// Don't pass a metrics registry, because config-reloader panics when we create our second.
+	tempReloader := newReloader(nil, &url.URL{
+		Scheme: "http",
+		// Since Prometheus is not ready, we won't be able to hit the reload URL so hit itself.
+		Host: listenAddr.String(),
```

Comment:
Yea, I would advocate to stop hacking like that (: Bit too much. 

All the reloader code is upstream. I technically maintain it with Thanos team. Either we fork/vendor that code or we fix there, this is too hacky.

On top of that literally nothing technically fails (as is - stops working) if you just run normal reloader as we have now (without creating new) even if Prometheus is down. The [reloading will retry, increment metric and log error.](https://github.com/thanos-io/thanos/blob/main/pkg/reloader/reloader.go#L534) - expected if Prometheus is slowly starting.

To me all we need to do is either:
1) Actually do what has been proposed 1y+ ago, [just remove readiness check](https://github.com/GoogleCloudPlatform/prometheus-engine/pull/526). Same arguments why it's not necessary apply now.
2) We could keep readiness check but start reloader early in the background if we want readiness crash (I don't think that's useful)
3) Change reloader upstream, so there is a lib function that only generates config, easy.
4) Embed config-reloader code to rule-eval.